src/java.base/share/classes/sun/security/provider/ML_DSA.java

```diff
@@ -0,0 +1,1427 @@
+     */
+
+    private void power2Round(int[][] input, int[][] lowPart, int[][] highPart) {
+        for (int i = 0; i < mlDsa_k; i++) {
+            for (int m = 0; m < mlDsa_n; m++) {
+                int rplus = input[i][m];
+                int r0 = input[i][m] & dMask;
+                int r00 = (1 << (mlDsa_d - 1)) - r0 ; // 2^d/2 - r+
+                r0 -= (r00 >> 31) & (1 << mlDsa_d); //0 if r+ < 2^d/2
+                lowPart[i][m] = r0;
+                highPart[i][m] = (rplus - r0) >> mlDsa_d;
+            }
+        }
+    }
+
+    private void decompose(int[][] input, int[][] lowPart, int[][] highPart) {
+        for (int i = 0; i < mlDsa_k; i++) {
+            for (int m = 0; m < mlDsa_n; m++) {
+                int rplus = (input[i][m] + mlDsa_q) % mlDsa_q;
+                int r0 = rplus % (2*gamma2);
+                r0 -= r0 > gamma2 ? 2*gamma2 : 0;
+                int r1;
+                if (rplus - r0 == mlDsa_q - 1) {
+                    r1 = 0;
+                    r0 = r0 - 1;
+                } else {
+                    r1 = (rplus - r0) / (2*gamma2);
+                }
+                lowPart[i][m] = r0;
+                highPart[i][m] = r1;
+            }
+        }
+    }
+
+    private int highBits(int input) {
+        int rplus = (input + mlDsa_q) % mlDsa_q;
+        int r0 = rplus % (2*gamma2);
+        r0 -= r0 > gamma2 ? 2*gamma2 : 0;
+        int r1;
+        if (rplus - r0 == mlDsa_q - 1) {
+            r1 = 0;
+            r0 = r0 - 1;
+        } else {
+            r1 = (rplus - r0) / (2*gamma2);
+        }
+        return r1;
+    }
+
+    private boolean[][] makeHint(int[][] z, int[][] r) {
+        int[][] v1 = vectorAdd(r,z);
+        boolean[][] res = new boolean[mlDsa_k][mlDsa_n];
+        for (int i = 0; i < mlDsa_k; i++) {
+            for (int j = 0; j < mlDsa_n; j++) {
+                int r1High = highBits(r[i][j]);
+                int v1High = highBits(v1[i][j]);
+                res[i][j] = r1High != v1High;
+            }
+        }
+        return res;
+    }
+
+    private int[][] useHint(boolean[][] h, int[][] r) {
+        int m = (mlDsa_q - 1) / (2*gamma2);
+        int[][] lowPart = new int[mlDsa_k][mlDsa_n];
+        int[][] highPart = new int[mlDsa_k][mlDsa_n];
+        decompose(r, lowPart, highPart);
+
+        for (int i = 0; i < mlDsa_k; i++) {
+            for (int j = 0; j < mlDsa_n; j++) {
+                if (h[i][j]) {
+                    highPart[i][j] += lowPart[i][j] > 0 ? 1 : -1;
+                }
+                highPart[i][j] = ((highPart[i][j] % m) + m) % m;
+            }
+        }
+        return highPart;
+    }
+
+    /*
+    NTT functions as specified in Section 7.5 of specification
+    */
+
+    public static int[] mlDsaNtt(int[] coeffs) {
+        int result = implMlDsaAlmostNtt(coeffs, montZetasForVectorNtt);
+        int[] check = coeffs.clone();
+        result = implMlDsaMontMulByConstant(coeffs,  montRModQ);
```

Comment:
In FIPS 204, NTT does not end with multiplying a constant. Why do you need one?

Reply:
That is the step that brings the result to the required range. The mod q computations are done using Montgomery multiplications and with the additions subtractions the results after each iteration of the loop may go outside the [-q, q] range, but this final step brings them back. It is basically a Montgomery multiplication by 1 and a conversion back from "Montgomery domain" to "normal domain", so the result does not change modulo q, and it will be in [-q, q].